### PR TITLE
Add specs for OAI-PMH

### PR DESCRIPTION
**Pull request by laritakr**
Completes DLTP-1758

### CODE REVIEW

File: spec/controllers/oai_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+  end
+
+  describe '#index' do
+    context 'with no params' do
+      let(:oai_params) {}
+      let(:key) { 'error' }
+      let(:attribute) { 'code' }
+      let(:value) { doc.css(key).attr(attribute).value }
+
+      it 'returns 200 with bad verb error' do
```
Comment (jeremyf) [inline spec/controllers/oai_controller_spec.rb:22]: This strikes me that a bad verb should NOT be a 200 status. However, if that is the OAI PMH spec, then I suppose we can proceed. Perhaps a `422 Unprocessable Entity` error?

File: spec/controllers/oai_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    end
+
+    context 'for verb GetRecord' do
+      context 'without required params' do
+        let(:oai_params) {{ verb: 'GetRecord' }}
+        let(:key) { 'error' }
+        let(:attribute) { 'code' }
+        let(:value) { doc.css(key).attr(attribute).value }
+
+        it 'returns 200 with bad argument error' do
```
Comment (jeremyf) [inline spec/controllers/oai_controller_spec.rb:47]: Again should this return a 422 status?

File: spec/controllers/oai_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        end
+      end
+
+      context 'with bad identifier' do
+        let(:oai_params) { { verb: 'GetRecord', identifier: 'i_dont_exist', metadataPrefix: 'oai_dc' } }
+        let(:key) { 'error' }
+        let(:attribute) { 'code' }
+        let(:value) { doc.css(key).attr(attribute).value }
+
+        it 'returns 200 with ' do
```
Comment (jeremyf) [inline spec/controllers/oai_controller_spec.rb:73]: This strikes me as a 404 error?

File: spec/controllers/oai_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    end
+
+    context 'for verb ListRecords' do
+      context 'without required parameters' do
+        let(:oai_params) {{ verb: 'ListRecords' }}
+        let(:key) { 'error' }
+        let(:attribute) { 'code' }
+        let(:value) { doc.css(key).attr(attribute).value }
+
+        it 'returns 200 with bad argument error' do
```
Comment (jeremyf) [inline spec/controllers/oai_controller_spec.rb:120]: Perhaps a 422 HTTP status?

File: spec/controllers/oai_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+        it 'returns 200 with the first page of results and a token to next page' do
+          get :index, oai_params
+          expect(value).to eq(1)
+          expect(token).to be_a(String)
+        end
+      end
+
+      context 'with date filtering for only past records' do
+        let(:oai_params) {{ verb: 'ListRecords', metadataPrefix: 'oai_dc', until: Time.now-2.days }}
+        let(:key) { 'error' }
```
Comment (jeremyf) [inline spec/controllers/oai_controller_spec.rb:180]: Is this key required?

File: spec/controllers/oai_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+          expect(value).to eq(1)
+          expect(token).to be_a(String)
+        end
+      end
+
+      context 'with invalid model set filter value' do
+        let(:oai_params) {{ verb: 'ListRecords', metadataPrefix: 'oai_dc', set: 'model:FakeModel' }}
+        let(:key) { 'error' }
+        let(:attribute) { 'code' }
+        let(:value) { doc.css(key).attr(attribute).value }
```
Comment (jeremyf) [inline spec/controllers/oai_controller_spec.rb:207]: Perhaps a 422?

**Comment by laritakr** [comment]
These errors are part of OAIPMH specs and come from the OAI gem. And the minimum required keys are the ones I included in the tests.

OAI standards are different from HTTP standards. See https://indico.cern.ch/event/5710/sessions/108048/attachments/988151/1405129/Simeon_tutorial.pdf for a simple explanation. Page 23, I believe, summarizes it this way:

> **OAI­PMH error handling**
> * all OK at HTTP level? => 200 OK
> * something wrong at OAI­PMH level? => OAI­PMH error (e.g. badVerb)

**Pull request merged**